API design: Self-service: Install Apple App Store apps on macOS (#22102)
fleetdm/fleet: docs/Contributing/API-for-contributors.md

- [Renew VPP token](#renew-vpp-token)
- [Delete VPP token](#delete-vpp-token)
- [Batch-apply MDM custom settings](#batch-apply-mdm-custom-settings)
- [Batch-apply packages](#batch-apply-packages)
- [Batch-apply App Store apps](#batch-apply-app-store-apps)
- [Get token to download package](#get-token-to-download-package)

> - [Download package using a token](#download-package-using-a-token)

Reviewer: @lucasmrod when we merged in reference docs for the 4.57 release, it looks like the batch-apply endpoints got moved to the REST API docs instead of the contributors docs.

I moved them back. When you get the chance, can you please check to see if these changes look right?

Also, I moved the download endpoints here b/c I think they're only for the Fleet UI to use (#21341). The best practice download endpoint for the IT admin to use in automations is here: https://fleetdm.com/docs/rest-api/rest-api#download-package

Reviewer 2: I'm super confused. It seems we deliberately moved them from API for contributors to REST API? ([here](https://github.com/fleetdm/fleet/pull/21912)).

Reviewer: Hmm, @rachaelshaw re your comment [here](https://github.com/fleetdm/fleet/pull/21912#issuecomment-2361363765), are we changing out philosophy? That is, endpoints that fleetctl uses live in API for contributors.

Reviewer 3: @noahtalerman nope, I was just confused; I saw Luke's review comment about the other `/app_store_apps` endpoints being the REST API and was like "oh yeah this looks like it fits with those" — I should have double-checked how this endpoint was being used first!

Reviewer: Got it!

> It seems we deliberately moved them from API for contributors to REST API?

@lucasmrod just following up here to say we goofed (which happens). We want the batch-apply endpoints to live in the API for contributors docs.

Reviewer: @rachaelshaw I think we're ready for final review/merge.

| scripts                                   | list   | body  | A list of script files to add to this team so they can be executed at a later time.                                                                                                                                                 |
| software                                   | object   | body  | The team's software that will be available for install.  |
| software.packages                          | list   | body  | An array of objects. Each object consists of:`url`- URL to the software package (PKG, MSI, EXE or DEB),`install_script` - command that Fleet runs to install software, `pre_install_query` - condition query that determines if the install will proceed, `post_install_script` - script that runs after software install, and `self_service` boolean.   |

> | software.app_store_apps                   | list   | body  | An array of objects. Each object consists of `app_store_id` - ID of the App Store app and `self_service` boolean. |

Reviewer 3: No need to make any changes now, but when we bring this into the regular API docs, that'd be a good time to update the format of this section so we're listing objects in separate tables (like the modify config docs)

  "exit_code": 0
}
```
## Software

### Batch-apply software

_Available in Fleet Premium._

`POST /api/v1/fleet/software/batch`

> This endpoint is asynchronous, meaning it will start a background process to download and apply the software and return a `request_uuid` in the JSON response that can be used to query the status of the batch-apply (using the `GET /api/v1/fleet/software/batch/{request_uuid}` endpoint defined below).

Reviewer 3: ```suggestion
This endpoint is asynchronous, meaning it will start a background process to download and apply the software and return a `request_uuid` in the JSON response that can be used to query the status of the batch-apply (using the `GET /api/v1/fleet/software/batch/:request_uuid` endpoint defined below).
```

| team_name | string | query | The name of the team to add the software package to. Ommitting these parameters will add software to 'No Team'. |
| dry_run   | bool   | query | If `true`, will validate the provided software packages and return any validation errors, but will not apply the changes.                                                                         |
| software  | object   | body  | The team's software that will be available for install.  |
| software.packages   | list   | body  | An array of objects. Each object consists of:`url`- URL to the software package (PKG, MSI, EXE or DEB),`install_script` - command that Fleet runs to install software, `pre_install_query` - condition query that determines if the install will proceed, `post_install_script` - script that runs after software install, and `uninstall_script` - command that Fleet runs to uninstall software. |

#### Example

`POST /api/v1/fleet/software/batch`

##### Default response

`Status: 200`
```json
{
  "request_uuid": "ec23c7b6-c336-4109-b89d-6afd859659b4",
}
```

### Get status of software batch-apply request

_Available in Fleet Premium._

> `GET /api/v1/fleet/software/batch/{request_uuid}`

Reviewer 3: ```suggestion
`GET /api/v1/fleet/software/batch/:request_uuid`
```

##### Default responses

`Status: 200`
```json
{
  "status": "processing",
  "message": "",
  "packages": null
}
```

`Status: 200`
```json
{
  "status": "completed",
  "message": "",
  "packages": [
    {
      "team_id": 1,
      "title_id": 2751,
      "url": "https://ftp.mozilla.org/pub/firefox/releases/129.0.2/win64/en-US/Firefox%20Setup%20129.0.2.msi"
    }
  ]
}
```

`Status: 200`
```json
{
  "status": "failed",
  "message": "validation failed: software.url Couldn't edit software. URL (\"https://foobar.does.not.exist.com\") returned \"Not Found\". Please make sure that URLs are reachable from your Fleet server.",
  "packages": null
}
```

### Batch-apply App Store apps

_Available in Fleet Premium._

`POST /api/latest/fleet/software/app_store_apps/batch`

#### Parameters

| Name      | Type   | In    | Description                                                                                                                                                           |
| --------- | ------ | ----- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | team_name | string | query | The name of the team to add the software package to. Ommitting this parameter will add software to 'No Team'. |

Reviewer 3: ```suggestion
| team_name | string | query | The name of the team to add the software package to. Ommitting this parameter will add software to "No team". |
```
